include/deal.II/numerics/kdtree_distance.h

```diff
@@ -0,0 +1,266 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef _dealii__numerics_kdtree_distance_h
+#define _dealii__numerics_kdtree_distance_h
+
+#include <deal.II/base/config.h>
+
+#  ifdef DEAL_II_WITH_NANOFLANN
+
+#include <deal.II/base/point.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <nanoflann.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the nanoflann library, used to compute the distance from a
+ * collection of points, and to efficiently return nearest neighbors to a
+ * target point. This function uses nanoflann to efficiently partition the
+ * space in a tree. The cost of each query is then roughly of order log(n),
+ * where n is the number of points stored in this class.
+ *
+ * The wrapper provides methods that give access to some of thefunctionalities
+ * of the nanoflann library, like searching the n nearest neighbors, or
+ * searching the points that fall within a raidius of a target point.
+ */
+template<int dim>
+class KDTreeDistance
+{
+public:
+  /**
+   * The max leaf parameter is used to decide how many points per leaf
+   * are used in the kdtree algorithm.
+   *
+   * If the points are not passed to this constructor, then you have
+   * to pass them later to this object by calling the set_points()
+   * method.
+   *
+   * Access to any of the methods without first passing a reference to
+   * a vector of points will result in an exception. Only a reference
+   * to the points is stored, so you should make sure that the life of
+   * the the vector you pass is longer than the life of this class, or
+   * you'll get undefinite behaviour.
+   *
+   * If you update the vector of points in someway, remember to call
+   * again the set_points() method. The tree and the index are
+   * constructed only once, when you pass the points (either at
+   * construction time, or when you call set_points()). If you update
+   * your points, and do not call again set_points(), your results
+   * will likely be wrong.
+   */
+  KDTreeDistance(const unsigned int &max_leaf_size=10,
+                 const std::vector<Point<dim> > &pts=std::vector<Point<dim> >());
+
+
+  /**
+   * Adaptor class used internally by nanoflann. This class stores a reference
+   * to the vector of points, and generates some helper functions for
+   * nanoflann.
+   */
+  struct PointCloudAdaptor
+  {
+    /**
+     * A typedef used by nanoflann.
+     */
+    typedef double coord_t;
+
+
+    /**
+     * Reference to the vector of points from which we want to compute
+     * the distance.
+     */
+    const std::vector<Point<dim> > &points; //!< A const ref to the data set origin
+
+
+    /**
+     * The constrcutor needs the data set source.
```

Comment:
Also `data set source` is likely some internal name? You do not use it elsewhere in this documentation, maybe `The constructor needs the vector of points from which we want to build the tree`?